storage/namespace.go

```diff
@@ -941,11 +955,15 @@ func (n *dbNamespace) readableShardAtWithRLock(shardID uint32) (databaseShard, e
 
 func (n *dbNamespace) initShards(needBootstrap bool) {
 	n.Lock()
+	var indexFn namespaceIndexWriteFn
+	if n.reverseIndex != nil {
+		indexFn = n.reverseIndex.Write
+	}
 	shards := n.shardSet.AllIDs()
 	dbShards := make([]databaseShard, n.shardSet.Max()+1)
 	for _, shard := range shards {
 		dbShards[shard] = newDatabaseShard(n.metadata, shard, n.blockRetriever,
-			n.namespaceReaderMgr, n.increasingIndex, n.commitLogWriter, n.reverseIndex,
+			n.namespaceReaderMgr, n.increasingIndex, n.commitLogWriter, indexFn,
```

Comment:
Is this literally just a function? Might be easier to maintain in the future if its a narrow interface

Reply:
happy to rename, any suggestions?

Reply:
The namespace owns the full reverse index type (with all the write/read methods). The shard is only given a `databaseIndexWriteFn`, defined as:

```
type databaseIndexWriteFn func(
	namespace ident.ID,
	id ident.ID,
	tags ident.Tags,
	fns indexInsertLifecycleHooks,
) error
```

That's kinda what you were going for, yea?

Reply:
I mean something like

```golang
type databaseIndex interface {
    Write(
        namespace ident.ID,
	id ident.ID,
	tags ident.Tags,
	fns indexInsertLifecycleHooks,
    ) error
}
```

And then if in the future we ever need to add another method (for whatever reason) its easier

Reply:
done. also renamed the associated types to `reverseIndexWriter` and `reverseIndexWriterFn` to align better with the naming in shard.go

Reply:
done

Reply:
The query will be transparent to users of the index, the write method will hide it.